frontend/src/components/Map/Map.js

```diff
@@ -0,0 +1,467 @@
+        map.current.addLayer(
+          {
+            id: hoverId,
+            type: 'line',
+            source: sourceName,
+            'source-layer': sourceLayer,
+            paint: {
+              'line-color': ['to-color', paint.hover?.outline?.color ?? paint?.default?.outline?.color ?? '#000000'],
+              'line-width': paint.hover?.outline?.width ?? paint?.default?.outline?.width ?? 1.0,
+              'line-opacity': [
+                'case',
+                ['to-boolean', ['feature-state', 'hover']],
+                paint.hover?.outline?.opacity ?? paint?.default?.outline?.opacity ?? 1.0,
+                0.0,
+              ],
+            },
+          },
+          beforeId,
+        )
+
+        layerIds.push(hoverId)
+
+        initPopup({ foreignKey, metricKey, titleKey, layerId: fillsId, sourceLayer })
+      }
+
+      setLayerIds(layerIds)
+      setSourceIds(sourceIds)
+    })
+  }, [data, initPopup, mapLoaded, metricConfig])
+
+  const cleanMap = useCallback(() => {
+    setLayerIds(layerIds => {
+      layerIds.forEach(id => {
+        const layer = map.current.getLayer(id)
+        if (layer) {
+          map.current.removeLayer(id)
+        }
+      })
+      return []
+    })
+
+    setSourceIds(sourceIds => {
+      sourceIds.forEach(id => {
+        const source = map.current.getSource(id)
+        if (source) {
+          map.current.removeSource(id)
+        }
+      })
+      return []
+    })
+  }, [])
+
+  // effects
+  useEffect(() => {
+    updateMap()
+
+    return () => cleanMap()
+  }, [cleanMap, updateMap])
+
+  useEffect(() => {
+    // initialize map only once
+    if (map.current || !config) {
+      return
+    }
+
+    setSelectedMetric(config?.defaultMetric)
+
+    const { style, bounds, fitBoundsOptions } = config?.options || {}
+
+    map.current = new mapboxgl.Map({
+      container: mapContainerRef.current,
+      style,
+      bounds,
+      fitBoundsOptions,
+    })
+
+    map.current.on('load', () => {
+      setMapLoaded(true)
+    })
+
+    map.current.addControl(new mapboxgl.NavigationControl({ showCompass: false }), 'bottom-right')
+
+    return () => {
+      map.current?.remove?.()
+      map.current = null
+    }
+  }, [config])
+
+  useEffect(() => {
+    const newMetricConfig = config?.metrics?.find(item => item?.id === selectedMetric) ?? {}
+    const colorScheme = newMetricConfig?.layers?.[0]?.paint?.default?.fill?.colorScheme ?? []
+    setMetricConfig(newMetricConfig)
+    setColorScheme(colorScheme)
+    getData(newMetricConfig?.data?.url)
+  }, [config, getData, selectedMetric])
+
+  return (
+    <Box position={'absolute'} top={0} bottom={0} left={0} right={0}>
+      <GlobalStyles styles={mapGlobalStyles} />
+
+      <div ref={mapContainerRef} style={{ height: '100%', width: '100%' }} />
+      <div ref={popupContainerRef} />
+      <Portal container={popupContainerRef.current}>
+        <MapPopupContent {...popupData} />
+      </Portal>
+
```

Comment:
@justinanderson I think we don't need a portal as we can just put MapPopupContent inside popupContainerRef div:
```suggestion
      <div ref={popupContainerRef}>
        <MapPopupContent {...popupData} />
      </div>
```

Reply:
Thanks, you're right.